### PR TITLE
tests: Add ability to restrict which tests run

### LINKED ISSUE #142: Allow individual tests to be run
The test script needs to allow the range of tests run to be restricted by either distro, or image type (image or initrd).

### PR DESCRIPTION
**Pull request by jodh-intel**
Added options to the test script to allow:
    
- all tests for a specific distro (`--distro $distro`).
- all image tests (`--test-images-only`).
- all initrd tests (`--test-initrds-only`).
    
Any combination of the above is also permitted. By default, all tests will be run.

Also added new doc for tests and renamed the script to `test_images.sh` as the name is clearer.
    
Fixes #142.
    
Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

File: tests/image_creation.sh [hunk 1]
```diff
@@ -346,6 +390,49 @@ test_distro_alpine()
...
+get_distro_test_names()
+{
+	typeset -F | awk '{print $3}' |\
+		grep "^${test_func_prefix}" | sort
+}
+
+# Displays a list of distros which can be tested
+list_distros()
+{
+	local funcs
```
Comment (grahamwhaley) [inline tests/image_creation.sh]: is `funcs` used here? feels like it might be a leftover.

File: tests/image_creation.sh [hunk 2]
```diff
@@ -346,6 +390,49 @@ test_distro_alpine()
...
+	
+	local defined_func
+	defined_func=$(echo "$test_funcs" | grep "^${expected_func}$" || true)
+
+	if [ -z "$defined_func" ]
+	then
+		local distros
+
+		# make a comma-separated list
+		distros=$(list_distros | tr '\n' ',' | sed 's/,$//g')
```
Comment (grahamwhaley) [inline tests/image_creation.sh]: you could pull that out and add it as part of the help or as a param (`--list-distros`)

**Comment by grahamwhaley** [comment]
@klynnrif @intelkevinputnam ptal

**Comment by grahamwhaley** [comment]
@jodh-intel the CI seems to not like your file rename patch:
```
Using index info to reconstruct a base tree...
M	tests/image_creation.sh
Falling back to patching base and 3-way merge...
Auto-merging tests/image_creation.sh
CONFLICT (content): Merge conflict in tests/image_creation.sh
error: Failed to merge in the changes.
Patch failed at 0001 tests: Rename distro test functions
The copy of the patch that failed is found in: .git/rebase-apply/patch
```

Not sure if that is a git glitch thing or if the patch needs to be in a slightly different format etc.

File: tests/image_creation.sh (hunk 1 above)
Comment (jodh-intel) [reply tests/image_creation.sh]: Good catch - removed.

File: tests/image_creation.sh (hunk 2 above)
Comment (jodh-intel) [reply tests/image_creation.sh]: Not sure I follow - this PR already includes a new `--list` option which lists the distros.

**Comment by jodh-intel** [comment]
My original branch was conflicted with `master` - let's see what the CI does this time... ;)

File: tests/image_creation.sh (hunk 2 above)
Comment (grahamwhaley) [reply tests/image_creation.sh]: Ah, so it does - I missed that. np then.

**Comment by jodh-intel** [comment]
The CI seems unhappy. Added `set -x` as this works fine locally for me.

/me puts a coin in the slot and cranks the CI again...

**Comment by chavafg** [comment]
@jodh-intel, sadly we have seen issues with OBS since last week, seems like the repo could not be synced.

```
Failed to synchronize cache for repo 'home_katacontainers_release', disabling.
Error: Unable to find a match
```

File: tests/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+* [Run the osbuilder tests](#run-the-osbuilder-tests)
+* [Further information](#further-information)
+
+## Run the osbuilder tests
+
+osbuilder provides a test script that will create all images and initrds for
```
Comment (klynnrif) [inline tests/README.md]: Lines 6-7 suggested rewrite: 
osbuilder provides a test script that creates all images and initrds for all supported distributions and then tests them to ensure a Kata Container can

File: tests/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+all supported distributions and then test them to ensure a Kata Container can
+be created with each.
+
+The test script will install all required Kata components on the host system
```
Comment (klynnrif) [inline tests/README.md]: The test script installs all required Kata components on the host system

**Comment by jodh-intel** [comment]
Thanks @klynnrif - branch updated.

**Comment by jodh-intel** [comment]
Sigh. This is a mess:

- centos CI timed out.
- fedora CI failed due to OBS cache sync error.
- ubuntu CI passed all tests, but somehow failed (...!?!)

Restarting and sacrificing more rubber chickens that last time in case that helps...

**Comment by jodh-intel** [comment]
/me grinds teeth due to network timeouts and OBS borkage...

**Comment by jodh-intel** [comment]
Ping @klynnrif.

**Comment by jodh-intel** [comment]
build timeouts yet again. Restarting CIs...

**Comment by jodh-intel** [comment]
Still experiencing network awfulness. It's awful :cry:

**Comment by jodh-intel** [comment]
@chavafg / @jcvenegas - any ideas how we can resolve these issues? This thing just will not pass the CI...

**Comment by chavafg** [comment]
I think we could use the static-binaries that are published when there is a new release [1], instead of using OBS, wdyt?

[1] https://github.com/kata-containers/runtime/releases/tag/1.1.0

**Comment by jodh-intel** [comment]
Hi @chavafg - nice idea! I've raised https://github.com/kata-containers/tests/issues/559 to add that feature to the kata-manager script which the tests here use.

Meantime though, I guess we have to hope OBS fixes itself...

**Comment by chavafg** [comment]
Cool, thank you. If I have some time today, I'll work on this to unblock the CI

**Comment by grahamwhaley** [comment]
I've nudged the two failing CIs anyhow, to see if they win the coin-toss-of-networkyness this time or not.
Let's keep an eye on them, and if they fail, we can manually/visually assess, as I suspect we are good to merge without all the CIs for this one...

**Comment by jodh-intel** [comment]
Still failing. I'll leave it up to you @grahamwhaley and @jcvenegas to decide whether to merge.

**Comment by grahamwhaley** [comment]
afaict, the failures have nothing to do with the PR. As the 16.04 CI passes, I'm happy to merge this. @jodh-intel @jcvenegas @chavafg - let's keep the failures in the back of our minds just in case we are wrong and they come back to bite us...

**Pull request merged**